src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Services/TagHelperSemanticRangeVisitor.cs

```diff
@@ -19,7 +20,7 @@ internal class TagHelperSemanticRangeVisitor : SyntaxWalker
         private readonly RazorCodeDocument _razorCodeDocument;
         private readonly Range? _range;
 
-        private TagHelperSemanticRangeVisitor(RazorCodeDocument razorCodeDocument, Range? range)
+        private TagHelperSemanticRangeVisitor(RazorCodeDocument razorCodeDocument, Range? range, TextSpan? rangeAsTextSpan) : base(rangeAsTextSpan)
```

Comment:
Do we still need the original `range`?

Reply:
I _think_ we still want the original `range`, unless we want to convert all `Ranges` to `TextSpans` [here](https://github.com/dotnet/razor-tooling/blob/2a4b6dd3d3550e6d42de9fe247bab25f65955887/src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Services/TagHelperSemanticRangeVisitor.cs#L489) (which seems inefficient). Maybe @ryanbrandenburg can correct me if I'm wrong. The Razor compiler changes I made in dotnet/razor-compiler#26 made it so only nodes overlapping with the passed-in range are visited. However, within a node that overlaps, there might be sub-parts of the node that we still don't want to include if they don't fall into the range.

Reply:
Ahhhh I see interesting ya we can keep. Should add a comment though

Reply:
> there might be sub-parts of the node that we still don't want to include if they don't fall into the range.

Hmm, this is a complicated scenario. Let's imagine that we have `<SurveyPrompt Title="My Title">SomeContents go here</SurveyPrompt>` on a page but no using. We add the using for SurveyPrompt and then VS requests `new Range(0, 14)` that would mean that the `SurveyPrompt` in the start tag would be colored as a RazorComponent, but not the one in the end-tag, and not the parameter `Title`. However, each of those is their own node which gets visited, not a sub-section of a single node so unless we "atomized" certain parent nodes together individually it would be impossibly to "do the right thing" here. If VS and VSCode request the range of the whole document in the absence of edit support, or if it requests ranges simultaneously (and they return roughly simultaneously) it's a bit academic but worth considering.

Regardless, I would vote that if a node falls at least partially within the range, then ALL its sub-parts (which are not themselves visitable nodes) should be evaluated for coloration.

Reply:
I looked over the code again and I think it might just be more efficient to remove the original `range`. We might be including slightly more nodes than we need to, but it would prevent us from having to do the extra computation to see if nodes overlap and would also help with code readability.

> Regardless, I would vote that if a node falls at least partially within the range, then ALL its sub-parts (which are not themselves visitable nodes) should be evaluated for coloration.

So I think this is what the PR already does. If a node partially falls within a range, we always evaluate it and its non-child sub-parts, check to see which child nodes also fall in the range, and then evaluate the child nodes that have any range overlap. Or I misunderstanding the concern?